homeassistant/components/config/category_registry.py

```diff
@@ -0,0 +1,134 @@
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "config/category_registry/list",
+        vol.Required("scope"): str,
+    }
+)
+@callback
+def websocket_list_categories(
+    hass: HomeAssistant, connection: ActiveConnection, msg: dict[str, Any]
+) -> None:
+    """Handle list categories command."""
+    category_registry = cr.async_get(hass)
+    connection.send_result(
+        msg["id"],
+        [
+            _entry_dict(entry)
+            for entry in category_registry.async_list_categories(scope=msg["scope"])
+        ],
+    )
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "config/category_registry/create",
+        vol.Required("scope"): str,
+        vol.Required("name"): str,
+        vol.Optional("icon"): vol.Any(cv.icon, None),
+    }
+)
+@websocket_api.require_admin
+@callback
+def websocket_create_category(
+    hass: HomeAssistant, connection: ActiveConnection, msg: dict[str, Any]
+) -> None:
+    """Create category command."""
+    category_registry = cr.async_get(hass)
+
+    data = dict(msg)
+    data.pop("type")
+    data.pop("id")
+
+    try:
+        entry = category_registry.async_create(**data)
+    except ValueError as err:
+        connection.send_error(msg["id"], "invalid_info", str(err))
+    else:
+        connection.send_result(msg["id"], _entry_dict(entry))
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "config/category_registry/delete",
+        vol.Required("scope"): str,
+        vol.Required("category_id"): str,
+    }
+)
+@websocket_api.require_admin
+@callback
+def websocket_delete_category(
+    hass: HomeAssistant, connection: ActiveConnection, msg: dict[str, Any]
+) -> None:
+    """Delete category command."""
+    category_registry = cr.async_get(hass)
+
+    try:
+        category_registry.async_delete(
+            scope=msg["scope"], category_id=msg["category_id"]
+        )
+    except KeyError:
+        connection.send_error(msg["id"], "invalid_info", "Category ID doesn't exist")
+    else:
+        connection.send_result(msg["id"])
+
+
+@websocket_api.websocket_command(
+    {
+        vol.Required("type"): "config/category_registry/update",
+        vol.Required("scope"): str,
+        vol.Required("category_id"): str,
+        vol.Optional("name"): str,
+        vol.Optional("icon"): vol.Any(cv.icon, None),
+    }
+)
+@websocket_api.require_admin
+@callback
+def websocket_update_category(
+    hass: HomeAssistant, connection: ActiveConnection, msg: dict[str, Any]
+) -> None:
+    """Handle update category websocket command."""
+    category_registry = cr.async_get(hass)
+
+    data = dict(msg)
+    data.pop("type")
+    data.pop("id")
+
+    try:
+        entry = category_registry.async_update(**data)
+    except ValueError as err:
+        connection.send_error(msg["id"], "invalid_info", str(err))
+    except KeyError:
+        connection.send_error(msg["id"], "invalid_info", "Category ID doesn't exist")
```

Comment:
It could also be the scope that is missing. The error message isn't wrong, as the category entry for that category id wouldn't exist in that case either, but it's confusing if we also don't mention that the scope is missing.

Reply:
Well, this is a safeguard, which should (technically not happen). If the category exists, the scope would, too (as a category is nested in the scope).

From the end-user, they can create categories, they can't create scopes. So, from that perspective, I think the current messaging is better in that regard.